src/framework.js

```diff
@@ -0,0 +1,194 @@
+/**
+ * File: framework.js 
+ */
+var _ = require('underscore');
+var util = require('util');
+
+var stateFactory = require('./create-state');
+var vnis = require('./vnis');
+
+/**
+ * This module implements the core of the RDF Pipeline Framework.  
+ */
+module.exports = {
+
+    componentFactory: function( nodeDef, wrapper ) { 
+        
+        var compFactory = require('./noflo-component-factory');
+        return compFactory(
+                _.defaults( {
+                     // Add an ondata function to be invoked every time we receive new data
+                     inPorts: _.mapObject(nodeDef.inPorts, function( port, portName ) {  
+                         return _.extend({
+                             ondata: function(payload, socketIndex) {
+                                   var that = this; 
+                                   return ondata.call( that, payload, portName, socketIndex, port );
+                             }
+                         }, port);
+                     })
+                 }, nodeDef ),
+                { rpf: {      // extend default noflo component node instance with framework rpf object 
+                     vnis: vnis,
+                     wrapper: wrapper
+                  }
+                } 
+        );
+    }
+
+}; // module.exports
+
+
+var ondata = function( payload, portName, socketIndex, port ) {
+
+     // socketIndex is only defined for addressable/multi ports so set default of 0 if not defined.  
+     // If there is only one input, we can just default to socket index 0, so if not defined, 
+     // do this for now.   We need to discuss whether addressable should just be required on our 
+     // pipeline ports.  IMHO, I think it probably should be to minimize potential user-error in this area
+     // and simplify the interface our users work with.
+     socketIndex = socketIndex || 0;
+
+     var vnid = payload.vnid || '';
+     var vni = vnis.get.call( this.nodeInstance, vnid );
+     vni.inputStates.set.call( vni.node, 
+                               portInfo.call( vni.node, portName, socketIndex ),
+                               stateFactory( vnid, payload ) );
+
+     // check if should run updater & run if necessary
+     if ( shouldRunUpdater( vni ) ) { 
+        var data = this.nodeInstance.rpf.wrapper.fRunUpdater( vni.inputStates );
+        if ( data ) {
+            var outputState = stateFactory( vnid, data );
```

Comment:
This should just set the data property of the existing output state object, instead of creating a new state object.

Also, it should generate and set a new LM in the state.